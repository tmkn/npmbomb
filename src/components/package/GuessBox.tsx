/** @jsx jsx */
import { jsx, css } from "@emotion/core";
import React, { useRef, useState, useContext } from "react";

import { PrimaryButton } from "../shared/buttons/Buttons";
import { mq, serifFont, primaryColorDark } from "../../css";
import { IPackageInfo } from "./Package";

export interface IGuessContext {
    package: IPackageInfo;
    guess: number | undefined;
    setUserGuess: (guess: number) => void;
}

export const GuessContext = React.createContext<IGuessContext>({
    package: {
        name: "",
        version: "",
        dependencies: 0,
        distinctDependencies: 0,
        directDependencies: 0,
        description: "",
        dependencyTree: {
            dependencies: [],
            name: ``,
            version: ``,
            transitiveCount: 0,
            loop: false
        }
    },
    guess: undefined,
    setUserGuess: () => {}
});

interface IGuessBoxProps {
    value?: string;
}

const guessBoxStyle = css({
    [mq[0]]: {
        display: "flex",
        flexDirection: "row",
        justifyContent: "center"
    }
});

const inputStyle = css({
    [mq[0]]: {
        padding: "0 1rem",
        fontSize: "1rem",
        fontFamily: `"${serifFont}"`,
        maxWidth: "5rem",
        border: "1px solid #B39DDB",
        outline: "none",
        color: `${primaryColorDark}`,
        textAlign: "center"
    }
});

export const GuessBox: React.FC<IGuessBoxProps> = ({ value }) => {
    const initValue: string = value ?? "";
    const [guess, setGuess] = useState(initValue);
    const [valid, setValid] = useState(Number.isNaN(parseInt(initValue)));
    const { setUserGuess } = useContext(GuessContext);

    function validate(e: React.ChangeEvent<HTMLInputElement>): void {
        const number = parseInt(e.target.value);

        setValid(Number.isNaN(number));
        if (Number.isNaN(number)) {
            setGuess("");
        } else {
            setGuess(number.toString());
        }
    }

    function doConfirm(): void {
        const number = parseInt(guess);

        if (!Number.isNaN(number)) {
            setUserGuess(number);
        }
    }

    return (
        <div css={guessBoxStyle}>
            <input
                css={inputStyle}
                value={guess}
                type="text"
                pattern="[0-9]"
                inputMode="numeric"
                onChange={validate}
                autoFocus
            />
            <div style={{ flex: "0.05" }} />
            <PrimaryButton disabled={valid} onClick={doConfirm}>
                Guess
            </PrimaryButton>
        </div>
    );
};
