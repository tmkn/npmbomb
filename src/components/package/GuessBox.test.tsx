import React from "react";
import { render, fireEvent, waitForElement } from "@testing-library/react";

import { GuessBox, IGuessContext, GuessContext } from "./GuessBox";

describe("<GuessBox />", () => {
    test("matches snapshot", () => {
        const { asFragment } = render(<GuessBox />);

        expect(asFragment()).toMatchSnapshot();
    });

    test("do valid input", async () => {
        const guess = "1337";
        const { container } = render(<GuessBox />);
        const input = await waitForElement(() =>
            container.querySelector<HTMLInputElement>("input[type='text']")
        );

        fireEvent.change(input!, { target: { value: guess } });

        expect(input!.value).toBe(guess);
    });

    test("do invalid input", async () => {
        const { container } = render(<GuessBox />);
        const input = await waitForElement(() =>
            container.querySelector<HTMLInputElement>("input[type='text']")
        );

        fireEvent.change(input!, { target: { value: "abc" } });

        expect(input!.value).toBe("");
    });

    test("executes guess callback", async () => {
        const callback = jest.fn();
        const mockContext: IGuessContext = {
            guess: undefined,
            package: null!,
            setUserGuess: callback
        };
        const { container, getByText } = render(
            <GuessContext.Provider value={mockContext}>
                <GuessBox />
            </GuessContext.Provider>
        );
        const input = await waitForElement(() =>
            container.querySelector<HTMLInputElement>("input[type='text']")
        );
        const btn = await waitForElement(() => getByText("Guess"));

        fireEvent.change(input!, { target: { value: "1337" } });
        fireEvent.click(btn);

        expect(callback).toHaveBeenCalled();
    });
});
