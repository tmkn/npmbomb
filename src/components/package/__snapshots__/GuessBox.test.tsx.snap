// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GuessBox /> matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="css-16fnzg4"
  >
    <input
      class="css-rznvji"
      inputmode="numeric"
      pattern="[0-9]"
      type="text"
      value=""
    />
    <div
      style="flex: 0.05;"
    />
    <button
      class="css-1w798tj"
    >
      Guess
    </button>
  </div>
</DocumentFragment>
`;
